### ISSUE TITLE
pathreparam integrator produces NaNs during rendering

### ISSUE THREAD

**Issue by mworchel** [mitsuba-renderer/mitsuba2#125]
I've built a small sample that uses the work in progress `pathreparam` integrator to optimize the vertex positions of a regular grid. The setup is pretty simple:

1. One camera looking in negative z direction from its origin (0, 0, 3)
2. A reference grid with vertices evenly distributed in xy directions and displaced along the z direction. 
3. A flat grid in the xy plane whos vertex displacements along the z direction will be optimized to match that of the reference grid.

The target image of the reference grid is rendered with the usual `path` integrator and the images of the grid to optimize are rendered with the `pathreparam` integrator (essentially following the `optim_vertices` sample in this [test repo](https://github.com/loubetg/mitsuba2-reparam-tests)). 

I noticed that the `pathreparam` integrator produces images that contain NaNs (more or less frequently depending on various parameters like learning rate, samples per pixels, etc.) which results in the loss becoming NaN. I suppose it's a bug in the integrator but it could be an issue with my sample of course.

I know that the integrator is currently work in progress and I'm not sure about its current state but I thought it might be interesting for you to know.

I attached the relevant script and data to reproduce the issue (for me, the loss becomes NaN in the second iteration). If no NaNs are produced, a sure way to provoke it, is making the `roughplastic` BSDF more smooth by lowering `alpha` from 0.1 to 0.01.

[displacement.zip](https://github.com/mitsuba-renderer/mitsuba2/files/4582253/displacement.zip)

**Comment by Speierers** [comment]
Hi @mworchel ,

Thanks a lot for reporting this. I will have a look at your script and try to debug this!

**Comment by Speierers** [comment]
Hi @mworchel ,

I just pushed a fix to the `pathreparam` branch (90b2510), could you please let me know if this fixes your issue?

Thanks!

**Comment by mworchel** [comment]
Hi @Speierers,

thank you for looking into it. I tried the fix and after playing around a bit I can say that it seems to behave much more stable now for various parameterizations.

However, I found that if you reduce the roughness of the material in the sample scene I provided from 0.1 to 0.01, one is still able to reliably produce NaNs. It seems like strong gradients and high learning rate drive parts of the mesh to some regions outside of the visible volume and I'm not completely sure what happens afterwards.

Here are the first three images of the optimization. The third image contains NaN values:
![out_000](https://user-images.githubusercontent.com/9449513/83162978-bf553a00-a10a-11ea-9c3e-89dc51e82a66.png) ![out_001](https://user-images.githubusercontent.com/9449513/83162998-c54b1b00-a10a-11ea-935b-862c7d4c2aee.png) ![out_002](https://user-images.githubusercontent.com/9449513/83163009-c9773880-a10a-11ea-86f0-e4a62a7ae83a.png)

Hope that helps!

**Comment by Speierers** [comment]
You need to be very careful when optimizing for the mesh parameters (e.g. vertex positions) directly. Nothing prevents the optimizer to create a degenerated mesh, with overlapping or collapsed triangles. I am not surprise you get NaNs in this case. They most likely come from the mesh itself rather than the integrator (e.g. normals computation, surface area computation, ...).

I would recommend to use a much smaller learning and somehow constraint the modification the optimizer makes to the vertex positions so that you always end up with a "valid" mesh a every iteration (e.g. only displace vertex position along one axis).

I will close this issue as this is most likely not a bug.

Thanks!

**Comment by mworchel** [comment]
Hi @Speierers ,

as stated in the OP, the vertex positions *are* already only displaced along *one* axis (the z-axis to be specifically). So the mesh does not contain any self intersections or degenerate triangles. That is exactly the reason, why I'm so suspicious about the NaNs.

**Comment by mworchel** [comment]
I dug a little deeper and I found that for the `roughness/alpha=0.01` scenario mentioned above the first NaNs appear in the displacement gradients in the *second* iteration. So the third image containing NaNs is just a consequence of this.

So basically we got the initial image (left in the comment above), then one optimization step that succeeds which results in the middle image and then follows the second optimization step (`ek.backward(loss); opt.step()`) that produces NaN gradients. The `loss` is a valid value, so the NaNs are produced somewhere during backpropagation.

I'm not sure if this is specifically linked to the pathreparam integrator or if its part of some other issue.

**Comment by Speierers** [comment]
Hi @mworchel ,

Could you have a look at the `pathreparam-optix7` branch and see if you still run into this issue? Take a look at the example script in `docs/examples/10_inverse_rendering/invert_heightfield.py`, the way you assign new values to the vertex positions have changed a bit.

**Comment by mworchel** [comment]
Hi, 

thanks for the update. I will try to adapt my sample and get back at you when I've news to share.
